### PR TITLE
OpenTelemetry Protocol with Apache Arrow Exporter initial skeleton

### PR DESCRIPTION
**Pull request by jmacd**
**Description:** First PR to introduce the OpenTelemetry Protocol with Apache Arrow exporter. From the upstream repository: https://github.com/open-telemetry/otel-arrow/blob/main/collector/exporter/otelarrowexporter

**Link to tracking Issue:**  #26491 

**Testing:** 
This is a skeleton PR, therefore only the skeleton contains tests.  Compared with the upstream repository, the `factory_test.go` and `config_test.go` files have been kept, the implementation tests in `otelarrow_test.go` were excluded in this PR.

**Documentation:** New README, [copied from the upstream repository](https://github.com/open-telemetry/otel-arrow/blob/main/collector/exporter/otelarrowexporter/README.md).

### CODE REVIEW

File: exporter/otelarrowexporter/go.mod [hunk 1]
```diff
@@ -0,0 +1,92 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/exporter/otelarrowexporter
+
+go 1.20
+
+require (
+	github.com/open-telemetry/otel-arrow v0.16.0
+	github.com/open-telemetry/otel-arrow/collector v0.16.0
```
Comment (codeboten) [inline exporter/otelarrowexporter/go.mod:7]: these dependencies will eventually go away once the implementation is complete in this repo correct?

File: exporter/otelarrowexporter/metadata.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+type: otelarrow
+
+status:
+  class: exporter
+  stability:
+    development: [traces, metrics, logs]
+  distributions: [contrib]
+  codeowners:
+    active: [jmacd, moh-osman3]
```
Comment (codeboten) [inline exporter/otelarrowexporter/metadata.yaml:9]: this doesn't match the code owners listed here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30619/files#diff-3d36a1bf06148bc6ba1ce2ed3d19de32ea708d955fed212c0d27c536f0bd4da7R71

File: exporter/otelarrowexporter/internal/arrow/exporter.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+package arrow // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/otelarrowexporter/internal/arrow"
+
+import (
+	"context"
+
+	arrowpb "github.com/open-telemetry/otel-arrow/api/experimental/arrow/v1"
+	"google.golang.org/grpc"
+)
+
+// Exporter is 1:1 with exporter, isolates arrow-specific functionality.
```
Comment (codeboten) [inline exporter/otelarrowexporter/internal/arrow/exporter.go:13]: Not sure i understand the comment here

File: exporter/otelarrowexporter/internal/arrow/exporter.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+// AnyStreamClient is the interface supported by all Arrow streams,
+// i.e., any of the Arrow-supported signals having a single method w/
+// the appropriate per-signal name.
+type AnyStreamClient interface {
+	Send(*arrowpb.BatchArrowRecords) error
+	Recv() (*arrowpb.BatchStatus, error)
+	grpc.ClientStream
+}
+
+// streamClientFunc is a constructor for AnyStreamClients.  These return
```
Comment (codeboten) [inline exporter/otelarrowexporter/internal/arrow/exporter.go:27]:
```suggestion
// StreamClientFunc is a constructor for AnyStreamClients.  These return
```

File: exporter/otelarrowexporter/config.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	// but we only support Zstd or none.
+	switch cfg.PayloadCompression {
+	case "none", "", configcompression.Zstd:
+	default:
+		return fmt.Errorf("unsupported payload compression: %s", cfg.PayloadCompression)
+	}
+	return nil
+}
+
+func (cfg *ArrowSettings) ToArrowProducerOptions() (arrowOpts []config.Option) {
```
Comment (codeboten) [inline exporter/otelarrowexporter/config.go:116]: Does this need to be public?

File: exporter/otelarrowexporter/README.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+Arrow](https://github.com/open-telemetry/otel-arrow) components with
+support for both OpenTelemetry Protocol with Apache
+Arrow and standard [OpenTelemetry Protocol](
+https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md)
+(OTLP) protocol via gRPC.
+
+## Getting Started
+
+The [OpenTelemetry Protocol with Apache
+Arrow](https://github.com/open-telemetry/otel-arrow) exporter combines
```
Comment (codeboten) [inline exporter/otelarrowexporter/README.md:25]: Should this link be updated? I assume the repo linked here will eventually be deprecated?

File: exporter/otelarrowexporter/README.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+Collector [OTLP
+exporter](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/otlpexporter)
+component with additional support for the OpenTelemetry Protocol with
+Apache Arrow.
+
+OpenTelemetry Protocol with Apache Arrow supports column-oriented data
+transport using the Apache Arrow data format.  This component converts
+OTLP data into an optimized representation and then sends batches of
+data using Apache Arrow to encode the stream.  The OpenTelemetry
+Protocol with Apache Arrow receiver <!-- TODO add link when the
```
Comment (codeboten) [inline exporter/otelarrowexporter/README.md:36]: This TODO can be addressed now

**Review by codeboten — COMMENTED** [review]
Thanks @jmacd! This looks good, just a few comments

File: exporter/otelarrowexporter/config.go (hunk 5 above)
Comment (jmacd) [reply exporter/otelarrowexporter/config.go]: Fixed.

File: exporter/otelarrowexporter/go.mod (hunk 1 above)
Comment (jmacd) [reply exporter/otelarrowexporter/go.mod:7]: The `otel-arrow` module contains a large piece of code for translating between Arrow and `pdata` OTLP. This library probably has uses outside of the Collector and I think it should remain separate, at least until the Arrow proto definitions are accepted formally as OTLP protocol.  There are OTel-Go SDK reasons we might use that library outside of collector components, probably.

The `otel-arrow/collector` module contains a number of components that were useful for us during development of the primary exporter/receiver, and we will want to keep them. They could each move into the `-contrib` repository, but that will take some additional effort, so I expect the repository itself to remain.

In addition to those statements, there are three packages for which an answer is needed, these being shared by the exporter and receiver and things I would recommend _for the core collector itself_. 

The `otel-arrow/collector/sharedcomponent` package is a copy of the one from the core repo's internal/ package. I'm not sure why the core repo won't export that functionality, so I replicated it.

[The `otel-arrow/collector/netstats` package is a set of very useful metrics that the Arrow components use, but which are applicable to all components](https://github.com/open-telemetry/opentelemetry-collector/issues/6638) and would also  useful for the automatic benchmark harness.

The `otel-arrow/collector/compression` package is an implementation that supports multiple configurable Zstd compression levels on a per-exporter component basis. See https://github.com/mostynb/go-grpc-compression/issues/24.

I suspect you will agree that all three of these packages belong in either `-collector` or `-collector-contrib`.

File: exporter/otelarrowexporter/internal/arrow/exporter.go (hunk 3 above)
Comment (jmacd) [reply exporter/otelarrowexporter/internal/arrow/exporter.go]: Fixed.

File: exporter/otelarrowexporter/metadata.yaml (hunk 2 above)
Comment (jmacd) [reply exporter/otelarrowexporter/metadata.yaml:9]: Fixed.

**Review by codeboten — APPROVED** [review]
Thanks for responding to my comments @jmacd, looks good

**Pull request merged**